### ISSUE TITLE
Interpretation of SHAP Values away from the mean

### ISSUE THREAD

**Issue by nickpschmidt** [shap/shap#435]
Hello,

I have a question about using SHAP values if the threshold for usage (e.g., getting offered a loan based on a model score) is below the average value of default. Specifically, how do you interpret and use the Shapley values?  Patrick Hall at H2O  (@jphall663) suggested I contact you about this. 

For example, suppose you have a really simple model:
Pr(D=1) = f(FICO, Income)

Average Pr(D=1) = 10%, but only 3% are approved.  This type of situation is extremely common in lending.

For person i: y-hat = 6%, where SHAP_FICO = -3%, and SHAP_Income = -1% (implying that person i's FICO and Income are both better than the average, but not good enough to get them below the 3% hurdle).

I think most people's intuition is to say that Income should be the "primary driver" because it is "less good", relative to the average than FICO.  But I don't think that is right: what drives person-i from 3% up to 6% (i.e., the threshold for acceptance) could be different than what drove him/her from 10% down to 6%.  For example, if the model is nonlinear, I think that you could have a situation where this person's FICO drove them up 2% relative to the average accepted person, and Income drove them up 1% relative to the average accepted person.  In that case, you would want to pick FICO as the primary driver, which is the opposite of what one would think based on ranking from the average percent of defaults.  Am I right about that?  Do you have any thoughts on the right approach?

I would really appreciate any guidance or thoughts you have about any of this.  Thanks!

**Comment by slundberg** [comment]
Great questions! SHAP values measure the impact of features by how much they change conditional expectations. Features will impact the conditional expectation most if they have "surprising" (i.e. rare) values and are also used strongly by the model. "Surprising" is always relative to some prior reference population. In the example you gave the reference population is all applicants and has a default rate of 10%. For person-i I would say FICO was the most informative feature, informative in the sense that it differentiates person-i from the average applicant in the eyes of the model. But as a person applying for the loan perhaps you don't care how you differ from the average applicant, perhaps you care more about how you differ from the typical accepted applicant. If that's the case then the "background" distribution used to compute SHAP values should probably not be a training set of all applicants, but instead just the set of accepted applicants. For a linear model this would just shift all SHAP values by the difference in the means between accepted applicants and all applicants. But for non-linear models it could in theory have a more complex effect on the values.

I should note that changing the background set in TreeExplainer is only supported for the `feature_dependence="independent"` option at the moment due to some technical details.

I should also note that explanations that have a very specific purpose (like how do I change my application so that I will be accepted) can be tailored to that purpose. For example, I can just tell you a list of changes that get you accepted and also minimize the deviation from the reference population. Now for the case of lenders, maybe they don't actually want to be that specific, so as to minimize the "gaming" of their algorithms.

**Comment by nickpschmidt** [comment]
Scott – thanks for getting back to me so quickly.  I’ve been thinking about what you wrote regarding training on the accepted/rejected dataset, but that it required the use the independent feature option.  Given that would likely be a big shortcoming, I think I have come up with a way around it.  I’d like to get your take on my idea.
 
What do you think of this:  Using the truly accepted/declined data (i.e., 3% get accepted in my example above), you build a new model on that data (the label is acceptance, the features are the same features in the base model).  This is essentially a surrogate model.  It can be an xgb, random forest, or whatever.  Second, you calculate SHAP values using the surrogate model.  The advantage is that since you are using a complex model and the right features as the surrogate, you should be able to get very good accuracy and capture the underlying process of the original model, but with the mean shifted down to 3%.  If my intuition is right then the resulting SHAP values should be very close to the right answer.  But I may be missing some important points, and your thoughts would really be appreciated.

Thanks again for getting back to me – and thanks for putting together some great software.

**Comment by jphall663** [comment]
@nickpschmidt, @slundberg feel free to ignore me, but I have some question and potential hints. 

***

@nickpschmidt - why is `feature_dependence="independent"` a drawback? My understanding is even this approach to Shapley values has consistency and local accuracy benefits over standard linear model coefficients and tree-based information-gain or permutation variable importance? 

*** 

All of this complexity is one reason why I use monotonic XGBoost -- *always*. Just makes these things easier to reason about. Hopefully that's another helpful hint (maybe you were already doing this).

Personally I use the Pearson correlation with the target in the training data to choose the direction of monotonicity per-feature, but there are probably smarter, more subtle ways to do this. 

***

Another potential hint here is to look at Friedman's H-stat in the region of the data you are concerned about, say 0.3 < yhat <= 0.1,  to see if there are strong interactions at play here.

http://statweb.stanford.edu/~jhf/ftp/RuleFit.pdf
https://christophm.github.io/interpretable-ml-book/interaction.html

I would hope if you've used monotonicity constraints, verified them with PDP/ICE/ALE, and H-stat values are low - indicating low interaction - you can reason through this fairly similarly to how you would think about a linear model, where each Shapley phi_ij is notionally similar to each linear model beta_j*x_ij product. 

*** 

If interactions are strong in this region, you could always fall back to a perturbation approach - what changes can move this person down to the acceptance criteria. But of course, changing single variables alone will lead to consistency problems ... so it's like you need to do the Shapley approach, but instead of only leaving a variable and subsets **out**, you need to **change** the variable and different subsets and take the appropriate weighted average across all subsets. 

Sorry I'm just rambling now ... but this is all super interesting to me.

**Comment by nickpschmidt** [comment]
Thanks @jphall663 - If I could get everyone to use monotonic constraints I would die a happy man.  :)

Personally, I like Spearman correlation because it only tests for monotonicity, rather than linearity.  And given the number of observations I work with, Spearman is sufficiently powerful that I do not really have to worry about false negatives.  The other thing I'll do is just run a polynomial regression and chart the expected values over the range of X.  But then you get into the von Neumann territory of "With four parameters I can fit an elephant, and with five I can make him wiggle his trunk."

I was just pointed to an article, [https://arxiv.org/pdf/1702.05471.pdf](url) that uses a method to calculate correlations between variables that I did not know of, where you do not assume monotonicity (that's only one step of what looks like a cool paper, but the one I'm most interested in right now).  I'm working my way through the paper and the idea seems quite interesting.

**Comment by jphall663** [comment]
Sorry @nickpschmidt ... kept editing after I made that comment. Bad habit. See above if any of this is helpful. Thanks!

***

The way I calculate non-linear, non-monotonic "correlation" (dependency?, association?) is very simple: I just train a non-linear model between the two variables and see if they can predict each other. You can do this for each pair of features in a dataset and then range-standardize between the highest and lowest RMSE and subtract values from 1 so you get what looks like a typical correlation matrix. 

You can even build such tables locally, on a row-by-row basis, by looking at raw, absolute, or deviance residuals for each of the models at each point. 

(To me this is super important in fairness, and people fail to realize this: multiple different features can act as proxies for each other to varying degrees at different local points in a dataset!!! Again, constrain, constrain, constrain your model! B/c you really have very little hope of reasoning through a fully nonlinear scenario ... IMO anyway. I know I am preaching to the choir :) )

(Not sure if that makes smart people cringe or not ... again I'm rambling so I'll stop.)

**Comment by slundberg** [comment]
Very interesting conversation :) ...I'll just comment on what @nickpschmidt suggested at first: Yes, that makes sense, and if the surrogate (mimic) model is good enough then you would get what you wanted. But I don't think the independent twiddling of features is something that is necessarily bad. It does potentially allow you to evaluate the model on unrealistic combinations of features (a downside), but it comes with the benefit of using an "intervention" to detect which features the model really depends on (a positive). It would take a lot of time to dig into the tradeoff of how/if to share credit among correlated inputs, but as it relates to this question, I would feel better using `feature_dependence="independent"` than I would using a mimic model.

**Comment by nickpschmidt** [comment]
@slundberg, is there a source that I could read about the feature_dependence option?  I read the bit in your code about it, but I don't think I followed it sufficiently.  It may be a flaw of mine, but every time I read the word "independent" I assume "unlikely condition to hold".  During my copious amounts of free time, I plan to run it both ways on some simulated data that I put together.  If the results are interesting then I'll post them.

@jphall663, one thought that I had is that you could run a SUR with all of the variables stacked on the left-hand side, with the non-linear versions of the variables on the right-hand side.  Then you could compare the joint F-Test from the groups of non-linearities to test the proxy condition.  The nice thing about that is that the joint F-tests are independent of scale, so if the variables are not standardized you still have a comparable statistic, which I don't think you do if you are using the RMSE index.  That would be helpful if you are using binary or categorical variables in the model.

The general issue that I think we face with proxies though is that they can be created through multiple interactions, whereas this method only captures a one-on-one variable relationship.  One way you can test if a model is proxying for a protected class is to create the model, get the predictions, and then run the predictions against protected class indicators (whether binary or proportional).  If you find that the class coefficients are not significant then you have an indication that the model is not proxying for protected class status.  The problem is that if you do find significance, you cannot disprove proxying, because the significance of the coefficients can indicate either (1) that you have a proxy, or (2) that there are still class-correlated elements for which you cannot control.  The latter is generally more likely since class-correlated relationships are very strong and hard to completely identify.

Yesterday I came up with another idea for measuring proxies that I think is novel and might just work (though my first pass at it turned out to be totally wrong).  I won't write it out here until I see if it works in order to save me from excess embarrassment if it is wrong.

**Comment by slundberg** [comment]
@nickpschmidt Sorry for the slow reply. I lost a battle with email this last month during a deadline push. Unfortunately I don't have much written up on that front yet. I have some figures built, and I'll try and post/publish/share them when I get a chance to describe them properly and make sure I understand it all.

**Comment by EtienneT** [comment]
This question was answered in march, but I was wondering if there was any new paper or article explaining more in details how to use the feature_dependence option?  I think it was renamed to feature_perturbation="interventional" in issue #882.

I would be really interested in understanding more what this option allows you to do.  Maybe with real life data scenarios and how to interpret the results.

Thank you!

**Comment by slundberg** [comment]
We have a paper coming out in a month that discusses this a bit. But the summary is that `feature_perturbation="path_dependent"` allows us to use the coverage information in the model to not need a background training dataset. The `feature_perturbation="interventional"` exactly matches the type of perturbation you would get with interventional conditional expectations, but needs to run over a provided background dataset (of usually 100 random samples or so). In most cases you should use the interventional option since over time I have found it to be a bit better, and also a couple recent papers argue for it. The path_dependent is sometimes faster though so that makes it nice. See #882 for an more details.

In practice the takeaway is that insights from causal statistics suggest that the interventional option is the right choice in most situations. There are two reasons you might want to use path_dependent instead: (1) you can use the training dataset for the background distribution that is implicitly encoded in the model (instead of explicitly defining the background dataset) and (2) it runs faster in some cases.

**Comment by EtienneT** [comment]
Thank you for your quick response!  Will definitely read the paper when it comes out.

One thing I am having difficulty understanding is the background dataset and when you would use a different background dataset than your training set.  You mentioned previously in this thread an example where maybe you want to know what are the factors that differentiate an observation from the people who were accepted for a loan instead of comparing them to the whole population.  So instead of using the training set as a background dataset maybe you can use only the accepted people.  But then how do we interpret the SHAP values?

Instead of using a background dataset containing only the accepted observations, could we filter the calculated SHAP values (without a background dataset, so using path_dependent) to only the observations that were accepted, aggregate the SHAP values per columns and compare those with an individual to see what features differed on average?

I found this example notebook in your repo that helped me understand a bit more: [Explaining a simple OR function](https://github.com/slundberg/shap/blob/master/notebooks/tree_explainer/Explaining%20a%20simple%20OR%20function.ipynb) which uses a background dataset.

One final question concerning feature_perturbation="interventional".  From what I understand, when using this option when you do feature permutations to calculate shapeley values, you remove certain features during permutations and you use causal graph rules from causal stats (the do operator) instead? A bit like the [dowhy sampler library from Microsoft](https://github.com/microsoft/dowhy).  But this doesn't mean we can make causal claims about the effects the SHAP values represents right?

Thanks!
